akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ReceiveBuilder.scala

```diff
@@ -24,9 +26,16 @@ final class ReceiveBuilder[T] private (
     private var messageHandlers: List[ReceiveBuilder.Case[T, T]],
     private var signalHandlers: List[ReceiveBuilder.Case[T, Signal]]) {
 
-  import ReceiveBuilder.Case
+  import ReceiveBuilder._
 
-  def build(): Receive[T] = new BuiltReceive[T](messageHandlers.reverse, signalHandlers.reverse)
+  def build(): Receive[T] = {
+    // signal handlers will often be empty so optimize for that
+    val builtSignalHandlers =
+      if (signalHandlers.isEmpty) defaultSignalHandlers[T]
+      else (adapterExceptionSignalHandler[T] :: signalHandlers).reverse
```

Comment:
I'm not so convinced about this optimization. 

`defaultSignalHandlers` is a single element list (constructed with `adapterExceptionSignalHandler :: Nil`), if `signalHandlers` is empty, calling `adapterExceptionSignalHandler :: signalHandlers` will have the same effect and we don't need to allocate `_defaultSignalHandlers`. 

Are we doing that just because we want to avoid a `.reverse` on a single element list? reverse is not optimized for single element list, so it will return another instance. But then, in that case, we are simply trading one extra instance for another. 

Or am I missing something here?

Reply:
It's based on a guess that in a majority of cases there will be no signal handlers added for a Java abstract behavior, in which case they can all share a single list instance with the adapter-failure-handling rather than create an identical copy of it, so basically avoiding one list alloc (potentially two with the reverse unless that is optimized in List) per Java actor, not a big thing but as they say "many small alloc streams becomes a big river"...